aes-kwp/src/kek.rs

```diff
@@ -0,0 +1,321 @@
+
+        // 2) Wrapping
+
+        // 2.1) Initialize variables
+
+        // Set A to the AIV
+        let mut block = GenericArray::<u8, Aes::BlockSize>::default();
+        block[..IV_LEN / 2].copy_from_slice(&IV_PREFIX);
+        block[IV_LEN / 2..IV_LEN].copy_from_slice(&mli);
+
+        // If n is 1, the plaintext is encrypted as a single AES block
+        if n == 1 {
+            // 1) Append padding
+
+            // GenericArrays should be zero by default, but zeroize again to be sure
+            for i in data.len()..n * SEMIBLOCK_SIZE {
+                block[IV_LEN + i] = 0;
+            }
+
+            block[IV_LEN..IV_LEN + data.len()].copy_from_slice(data);
+
+            self.cipher.encrypt_block(&mut block);
+            out.copy_from_slice(&block);
+        } else {
+            // 1) Append padding
+
+            // Don't trust the caller to provide a zeroized out buffer, zeroize again to be sure
+            for i in data.len()..n * SEMIBLOCK_SIZE {
+                out[IV_LEN + i] = 0;
+            }
+
+            // The following code is identical to AES-KW
+
+            // 2.2) Calculate intermediate values
+            out[IV_LEN..IV_LEN + data.len()].copy_from_slice(data);
+
+            for j in 0..=5 {
+                for (i, chunk) in out[IV_LEN..].chunks_mut(8).enumerate() {
+                    // A | R[i]
+                    block[IV_LEN..].copy_from_slice(chunk);
+                    // B = AES(K, ..)
+                    self.cipher.encrypt_block(&mut block);
+
+                    // A = MSB(64, B) ^ t
+                    let t = (n * j + (i + 1)) as u64;
+                    for (ai, ti) in block[..IV_LEN].iter_mut().zip(&t.to_be_bytes()) {
+                        *ai ^= ti;
+                    }
+
+                    // R[i] = LSB(64, B)
+                    chunk.copy_from_slice(&block[IV_LEN..]);
+                }
+            }
+
+            // 2.3) Output the results
+            out[..IV_LEN].copy_from_slice(&block[..IV_LEN]);
+
+            // End of the identical code
+        }
+
+        Ok(())
+    }
+
+    /// Computes [`Self::wrap`], allocating a [`Vec`] for the return value.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    pub fn wrap_vec(&self, data: &[u8]) -> Result<Vec<u8>> {
+        let n = (data.len() + SEMIBLOCK_SIZE - 1) / SEMIBLOCK_SIZE;
+        let mut out = vec![0u8; n * SEMIBLOCK_SIZE + IV_LEN];
+        self.wrap(data, &mut out)?;
+        Ok(out)
+    }
+
+    /// AES Key Wrap with Padding, as defined in RFC 5649.
+    ///
+    /// The `out` buffer will be overwritten, and must be exactly [`IV_LEN`]
+    /// bytes (i.e. 8 bytes) shorter than the length of `data`.
+    /// This method returns the size of the original key; the caller should
+    /// use this size to remove the padding, by truncating `out` to the
+    /// appropriate length.
+    pub fn unwrap(&self, data: &[u8], out: &mut [u8]) -> Result<usize> {
```

Comment:
I think it's more idiomatic in Rust to return a slice here, rather than a `usize`:

```suggestion
    pub fn unwrap<'a>(&self, data: &[u8], out: &'a mut [u8]) -> Result<&'a [u8]> {
```

Reply:
Oh, I didn't know that was possible in Rust, does this also work in a `no-std` environment?

Reply:
Yep. All you need to do is slice the output buffer at the end.